src/reflect/scala/reflect/internal/Constants.scala

```diff
@@ -83,7 +83,26 @@ trait Constants extends api.Constants {
     // !!! In what circumstance could `equalHashValue == that.equalHashValue && tag != that.tag` be true?
     override def equals(other: Any): Boolean = other match {
       case that: Constant =>
-        this.tag == that.tag && equalHashValue == that.equalHashValue
+        this.tag == that.tag && {
+          //
+          // Consider two `NaN`s to be identical, despite non-equality
+          // Consider -0d to be distinct from 0d, despite equality
+          //
+          // We use the raw versions (i.e. `floatToRawIntBits` rather than `floatToIntBits`)
+          // to avoid treating different encodings of `NaN` as the same constant.
+          // You probably can't express different `NaN` varieties as compile time
+          // constants in regular Scala code, but it is conceivable that you could
+          // conjure them with a macro.
+          //
+          (this.value == null && that.value == null) || (this.tag match {
+            case FloatTag =>
+              floatToRawIntBits(value.asInstanceOf[Float]) == floatToRawIntBits(that.value.asInstanceOf[Float])
+            case DoubleTag =>
+              doubleToRawLongBits(value.asInstanceOf[Double]) == doubleToRawLongBits(that.value.asInstanceOf[Double])
+            case _ =>
+              this.value.equals(that.value)
```

Comment:
I pushed a change to make it clearer that this won't NPE

Reply:
👍